src/fallback/dom_parser_implementation.cpp

```diff
@@ -243,6 +243,93 @@ WARN_UNUSED error_code implementation::minify(const uint8_t *buf, size_t len, ui
   // for fear of aliasing
   return SUCCESS;
 }
+WARN_UNUSED bool implementation::validate_utf8(const char *buf, size_t len) const noexcept { 
+  const uint8_t *data = (const uint8_t *)buf;
+  uint64_t pos = 0;
+  uint64_t next_pos = 0;
+  uint32_t code_point = 0;
+  while (pos < len) {
+
+    // check of the next 8 bytes are ascii.
+    next_pos = pos + 16;
+    if (next_pos <=
+        len) { // if it is safe to read 8 more bytes, check that they are ascii
+      uint64_t v1;
+      memcpy(&v1, data + pos, sizeof(uint64_t));
+      uint64_t v2;
+      memcpy(&v2, data + pos + sizeof(uint64_t), sizeof(uint64_t));
+      uint64_t v{v1 | v2};
+      if ((v & 0x8080808080808080) == 0) {
+        pos = next_pos;
+        continue;
+      }
+    }
+    unsigned char byte = data[pos];
+
+    if (byte < 0b10000000) {
```

Comment:
I've concluded this looks good. We should subject the UTF-8 checkers specifically to more rigorous testing, though.

Reply:
I have added tests right now.